GWAS.pm

```diff
@@ -0,0 +1,510 @@
+          }
+        } split(/[;,x]/, $rs_risk_allele);
+        $t_data->{"GWAS_risk_allele"} = $risk_allele;
+
+        push(@phenotypes, $t_data);
+      } @ids;
+    }
+  }
+  close($input_FH);
+
+  my %result = ('phenotypes' => \@phenotypes);
+  return \%result;
+}
+
+sub create_curated_processed_file {
+  my ($self, $data) = @_;
+  
+  my $temp_processed_file = $self->{"processed_file"} . "_temp";
+  open(my $temp_processed_FH, '>', $temp_processed_file) || die ("Could not open " . $self->{processed_file} . " for writing: $!\n");
+  
+  foreach my $phenotype (@{ $data->{"phenotypes"} }){
+    foreach my $vf (@{ $phenotype->{"vfs"} }){
+      next unless $vf;
+      
+      next unless $phenotype->{"GWAS_risk_allele"};
+      my $GWAS_risk_allele = $phenotype->{"GWAS_risk_allele"} || "";
+      
+      my $GWAS_associated_gene = $phenotype->{"GWAS_associated_gene"} || "";
+      my $GWAS_p_value = $phenotype->{"GWAS_p_value"} || "";
+      my $GWAS_study = $phenotype->{"GWAS_study"} || "";
+      my $GWAS_pmid = $phenotype->{"GWAS_pmid"} || "";
+      my $accessions = join(",", @{ $phenotype->{"GWAS_accessions"} }) || "";
+      my $GWAS_beta_coef = $phenotype->{"GWAS_beta_coef"} || "";
+      my $GWAS_odds_ratio = $phenotype->{"GWAS_odds_ratio"} || "";
+
+      my $line = join("\t", (
+        $vf->{"seq"}, $vf->{"start"}, $vf->{"end"}, $vf->{"ref"}, 
+        $GWAS_associated_gene,
+        $GWAS_risk_allele,
+        $GWAS_p_value,
+        $GWAS_study,
+        $GWAS_pmid,
+        $accessions,
+        $GWAS_beta_coef,
+        $GWAS_odds_ratio,
+      ));
+      
+      print $temp_processed_FH $line . "\n";
+    }
+  }
+  
+  close($temp_processed_FH);
+  
+  system("sort -k1,1 -k2,2n -k3,3n $temp_processed_file | bgzip -c > $self->{processed_file}") == 0
+    or die "Failed to sort and compress $temp_processed_file: $?\n";
+  system("rm $temp_processed_file") == 0
+    or die "Failed to delete $temp_processed_file: $?\n";
+  system("tabix -s 1 -b 2 -e 3 -f " . $self->{processed_file}) == 0
+    or die "Failed to create index " . $self->{processed_file} . ": $?\n";;
+}
+
+sub parse_sstate_header {
+  my ($self) = @_;
+  
+  my $header = `head -n 1 $self->{"file"}` or die "Cannot get header from " . $self->{"file"} . ": $?\n";
+  $header =~ s/^#//;
+  
+  my @cols = (split("\t", $header));
+  my @required_cols = qw/hm_chrom hm_pos p_value hm_beta hm_odds_ratio hm_effect_allele/;
+  
+  my $colmap = {};
+  map {
+    my $index = $_;
+    my @matched = (grep {/^$cols[$index]$/} @required_cols);
+    $colmap->{$index} = $matched[0] if @matched;
+  } 0 .. $#cols;
+
+  return $colmap;
+}
+
+sub create_state_processed_file {
+  my ($self, $c, $p) = @_;
+  
+  my $file = $self->{"file"};
+  my $processed_file = $self->{"processed_file"};
+  
+  my $filtered_file = $file . ".filtered";
+  system("tail -n +2 $file | awk '\$$c != \"NA\" && \$$p != \"NA\" {print}' > $filtered_file") == 0
```

Comment:
Since most files will be compressed I think it would be nice to decompress first
```suggestion
  system("gunzip -cf $file | tail -n +2 $file | awk '\$$c != \"NA\" && \$$p != \"NA\" {print}' > $filtered_file") == 0
```